### ISSUE TITLE
File Auto Reloading with Untracked Git Files During Branch Change

### ISSUE THREAD

**Issue by acb** [atom/atom#9257]
Firstly:
Atom 1.0.19
Ubuntu 14.04.2 LTS

The file buffer autoloading fails to load changes from `git stash pop`s on files that didn't exist in the previous branch you were on.  That's kind of a confusing sentence so I'll elaborate.  In this repo https://github.com/acb/Atom-Test there is a branch `master` with the file `master.txt` in it.  There is also a branch `dev` that has a file `dev.txt` in it that does not exist in master.

Now perform the following steps (leaving `dev.txt` open in Atom the entire time):
1. `git checkout dev`
2. Alter `dev.txt` in some way
3. `git stash` -- changes in `dev.txt` are stashed and the buffer in Atom changes appropriately
4. `git checkout master`
5. `git checkout dev`
6. `git stash pop` -- changes in `dev.txt` are `pop`ped to disk, but the buffer in Atom **does not** update

No errors are shown in the console.

If this goes undected by the user they could accidentally save `dev.txt` and blow away their stashed changes without any warning.  The solution as of now is to close the file (and DON'T SAVE CHANGES when prompted) and reopen.

Attached is a gif of the behavior.  Please let me know if I can provide any more information, and feel free to use the repo linked above for any testing you may need.

![atom-bug](https://cloud.githubusercontent.com/assets/4690898/10678022/bcda2682-78dd-11e5-830e-c29b223c1763.gif)

**Comment by thedaniel** [comment]
Thanks for the detailed repro steps!

**Comment by thanhtuong** [comment]
I also got the same problem like that. Please help to fix! Before work with atom, I worked with sublime, and it very smooth!

**Comment by mrerichoffman** [comment]
Still an issue as of 1/29/16 -- I may leave atom editor due to bug. Can potentially overwrite a file without any indications from the editor. I have reproduced with git stash and checkout.

**Comment by acb** [comment]
I changed the title of the bug to highlight the potential for data loss.  Hopefully this gets addressed sooner than later, I've had a couple close calls where I almost lost my stashed changes.

**Comment by mrerichoffman** [comment]
Thanks hopefully this gets worked!

**Comment by chrwi** [comment]
Definitely related to https://github.com/atom/atom/issues/4982#issuecomment-195806200, and it has been noticed by the devs

**Comment by lbraiden-r7** [comment]
Just lost work because of this.  Also forced to leave atom due to this.  Please fix.

**Comment by caesarsol** [comment]
I get a similar problem when simply modifying any unsaved file in another editor, do you think it's the same cause?

1. Open `a.txt` in Atom
0. Open `a.txt` in another editor (I'm using Sublime Text)
0. Write `aaa` in Atom without saving
0. Write `bbb` in Sublime without saving - at this stage, the two unsaved files "diverge"
0. Save the buffer in Sublime
0. Save the buffer in Atom
0. The file contains `aaa`, **Atom has replaced its content without alerting you may be losing data**

The same steps replacing Atom and Sublime causes a Sublime alert:

```
The file 
a.txt 
Has changed on disk.
Do you want to reload it?
```

I think it's the same problem, do you think it is? Should I open another issue?

**Comment by chriswils** [comment]
I guess https://atom.io/packages/file-watcher should be in core. Would definitely fix this. Used it a long time now due to this issue

**Comment by farvardin** [comment]
yeah @caesarsol it's exactly the expected result. All other good txt editor I know (kate, geany, scite) are taking this into account.

@chriswils file-watcher is working good, yes it should be in core if it's not the case already.

**Comment by ungb** [comment]
I am able to repro this issue with the detailed repro steps @acb  provided.   


If you check `close deleted file tab` this would cause this issue to not occur, but it's an issue only when `Close Deleted File Tabs` is **not** check, but this is still an issue.

**Comment by ianformanek** [comment]
Just thought I lost half day of work due to this and spent 2 hours figuring out what got lost where and decompiling affected files and diff-ing the changes.

Ultimately found out that when Atom lets you down, git comes to the rescue - you can still recover dropped/applied stash - so I leave it here for the future unfortunate victims who may benefit from finding this faster: https://stackoverflow.com/questions/89332/how-to-recover-a-dropped-stash-in-git

**Comment by kuychaco** [comment]
Another example of how this issue manifests - https://github.com/atom/atom/issues/14765#issuecomment-307974909

**Comment by kuychaco** [comment]
The solution for this is blocked until @smashwilson's file change API is in place.

In the meantime I'll document my findings here to retain the context.

Repro steps:

**Scenario 1**
1. Create a branch `foo` with a file `foo.txt` that does not exist on `master`. Commit this file and keep `foo.txt` open in the pane.
2. Uncheck `Close Deleted File Tabs`
3. `git checkout master` - `foo.txt` should remain open but is now marked as a modified unsaved file
   * if `Close Deleted File Tabs` is checked, then the file should close itself
4. `git checkout foo` - `foo.txt` remains modified, whereas it should be marked as unmodified

Herein lies the core problem. We don't detect that `foo.txt` has been restored back on disk and take the buffer out of the modified state. Since we don't update modified buffers when disk contents change in order to avoid the loss of unsaved changes, the changes from popping the stash don't appear in the buffer. And saving the modified buffer overwrites the changes just popped off of the stack. 

There's another related data loss scenario that is slightly different and involves file modification as opposed to creation. 

**Scenario 2**
1. Make changes to `file.txt`, save and stash them
2. Make conflicting changes to `file.txt` and do not save them
3. Pop the original changes off the stash

Actual: the disk updates to have the popped changes, but the buffer doesn't change because it's in a modified state. If you were to save the file, the changes on disk would be overwritten.
Expected: the best UX is still to be determined (@simurai have any thoughts?). We could detect the changes on disk for the file that is in a modified state and then open a modal or show a banner that notifies the user that the file has been modified on disk and asks them if they'd like to keep their unsaved changes, discard them and use the disk contents, or perhaps preview the merged changes.

To sum up I think there are two things we need to do in order to handle this issue correctly:

1. Mark the buffer as unmodified when the disk contents are updated and match what's in the buffer, so that subsequent changes on disk are reflected in the buffer. 
    1. if the underlying file is going from _deleted_ to _created_ (scenario 1) we need to watch the folder for the creation of new files, compare the contents of the disk and the buffer, and remove the modified state if they match. To watch the folder for the creation of new files we need to wait for @smashwilson's new file change API to be in place. 
    2. if the underlying file is simply modified (scenario 2) then we need to detect when the underlying file has changed (should already be possible using the current pathwatcher api), and prompt the user for what they would like to do. A fix can already be made for this case, but first UX decisions need to be made.

**Comment by simurai** [comment]
> We could detect the changes on disk for the file that is in a modified state and then open a modal or show a banner that notifies the user that the file has been modified on disk and asks them if they'd like to keep their unsaved changes, discard them and use the disk contents, or perhaps preview the merged changes.

Here a modal that is only shown inside the editor. So it only starts to bother you once you want to close the file or make changes.

![unsaved-file](https://user-images.githubusercontent.com/378023/29393593-a6e155a8-833f-11e7-8dd7-aded034d286e.png)

Is the wording clear enough? It's almost like a "resolve merge conflict" and could show a diff. But probably out of scope for an initial version.

**Comment by lbraiden-r7** [comment]
"got changed" is poor grammar.  Should say "This file has been modified by another program" or similar, like most editors do.  Would suggest a little more explanation of the implications of each choice, too.  Perhaps rename "Continue editing" to "Keep this version", to make it more clear that it's a choice between which data to lose.

**Comment by binaryfunt** [comment]
I tried using the [file-watcher package](https://atom.io/packages/file-watcher) with "Prompt on change" ticked, but it didn't do anything at all when I edited a file, stashed, checked out another branch, checked out back again, popped the stash and started editing the non-updated file on buffer

**Comment by pingpong2011** [comment]
This issue still exists with file-watcher and worse. I don't get a prompt when it reloads a file from disk, and if Atom is not in focus, nothing happens at all, and since Atom can't reload a file from disk, I lose changes made _outside_ the program.

If I edit the document, even with Atom in focus it doesn't reload the file nor ask me if I want to preserve my current file vs reload the one from disk. I wanted to use Atom but this is ridiculous.

**Comment by chriswils** [comment]
My solution: Move to VS Code, with atom keymapping.

**Comment by baverhey** [comment]
Still an issue in september2019

**Comment by farvardin** [comment]
Isn't it fixed yet? Strange
